src/engines/bing.rs

```diff
@@ -0,0 +1,109 @@
+    parser: SearchResultParser,
+}
+
+impl Bing {
+    /// Creates the Bing parser.
+    pub fn new() -> Result<Self, EngineError> {
+        Ok(Self {
+            parser: SearchResultParser::new(
+                ".b_results",
+                ".b_algo",
+                "h2 a",
+                ".tpcn a.tilk",
+                ".b_caption p",
+            )?,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl SearchEngine for Bing {
+    async fn results(
+        &self,
+        query: &str,
+        page: u32,
+        user_agent: &str,
+        client: &Client,
+        _safe_search: u8,
+    ) -> Result<HashMap<String, SearchResult>, EngineError> {
+        // Bing uses `start results from this number` convention
+        // So, for 10 results per page, page 0 starts at 1, page 1
+        // starts at 11, and so on.
+        let results_per_page = 10;
+        let start_result = results_per_page * page + 1;
+
+        let url: String = match page {
+            0 => {
+                format!("https://www.bing.com/search?q={query}")
+            }
+            _ => {
+                format!("https://www.bing.com/search?q={query}&first={start_result}")
+            }
+        };
+
+        let header_map = HeaderMap::try_from(&HashMap::from([
+            ("USER_AGENT".to_string(), user_agent.to_string()),
+            ("REFERER".to_string(), "https://google.com/".to_string()),
+            (
+                "CONTENT_TYPE".to_string(),
+                "application/x-www-form-urlencoded".to_string(),
+            ),
+        ]))
+        .change_context(EngineError::UnexpectedError)?;
+
+        let document: Html = Html::parse_document(
+            &Bing::fetch_html_from_upstream(self, &url, header_map, client).await?,
+        );
+
+        // Bing is very aggressive in finding matches
+        // even with the most absurd of queries. ".b_algo" is the
+        // class for the list item of results
+        if let Some(no_result_msg) = self.parser.parse_for_no_results(&document).nth(0) {
+            if no_result_msg
+                .value()
+                .attr("class")
+                .map(|classes| classes.contains("b_algo"))
+                .unwrap_or(false)
+            {
+                return Err(Report::new(EngineError::EmptyResultSet));
+            }
+        }
+
+        let re_span = Regex::new(r#"<span.*?>.*?(?:</span>&nbsp;·|</span>)"#).unwrap();
+        let re_strong = Regex::new(r#"(<strong>|</strong>)"#).unwrap();
+
+        // scrape all the results from the html
+        self.parser
+            .parse_for_results(&document, |title, url, desc| {
+                Some(SearchResult::new(
+                    &re_strong.replace_all(title.inner_html().trim(), ""),
+                    url.value().attr("href").unwrap(),
+                    &re_strong.replace_all(&re_span.replace_all(desc.inner_html().trim(), ""), ""),
+                    &["bing"],
+                ))
+            })
+    }
+}
```

Comment:
```suggestion

        // scrape all the results from the html
        self.parser
            .parse_for_results(&document, |title, url, desc| {
                Some(SearchResult::new(
                    &re_strong.replace_all(title.inner_html().trim(), ""),
                    url.value().attr("href").unwrap(),
                    desc.inner_html().trim(),
                    &["bing"],
                ))
            })
    }
}
```

I think you don't need to replace `span` and `strong` tags because they will be rendered on the search engine page. Also, having this would also be necessary as this will highlight what matched in the description of the site because of which these results were shown. This can add a little aesthetic style which can provide a good user experience to the user. So I would suggest leaving it :slightly_smiling_face: .

Reply:
For the `strong` tags, I figured it was necessary to sanitize and to maintain uniformity among the search engines, i.e. other search engines do not insert semantic tags on their results. But if we want that in our results, I'll remove it

For the `span` tags, I believe that one is necessary to be removed because it removes a Bing-specific element not relevant to the results

### Screenshot from Bing. The _Web_ text is inside the span which I removed.

![2024-01-01_08-02](https://github.com/neon-mmd/websurfx/assets/31008330/ad4b1b36-6282-4ca8-b648-13e151cc462a)

### Screenshot from Websurfx with the `strong` and `span` tags included
 
![2024-01-01_08-03](https://github.com/neon-mmd/websurfx/assets/31008330/ae2bb279-804f-4085-9ada-4970c217d973)

Reply:
> For the `strong` tags, I figured it was necessary to sanitize and to maintain uniformity among the search engines, i.e. other search engines do not insert semantic tags on their results. But if we want that in our results, I'll remove it
> 
> For the `span` tags, I believe that one is necessary to be removed because it removes a Bing-specific element not relevant to the results
> 
> ### Screenshot from Bing. The _Web_ text is inside the span which I removed.
> ![2024-01-01_08-02](https://private-user-images.githubusercontent.com/31008330/293525761-ad4b1b36-6282-4ca8-b648-13e151cc462a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQwODI0MTYsIm5iZiI6MTcwNDA4MjExNiwicGF0aCI6Ii8zMTAwODMzMC8yOTM1MjU3NjEtYWQ0YjFiMzYtNjI4Mi00Y2E4LWI2NDgtMTNlMTUxY2M0NjJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAxVDA0MDgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3YTA2Y2E0MzhmM2FlNDE1YmVkNDg2YjI4NWFiN2EwNWM5ODI5MzQ4Y2RhODA3OGQ1N2I5ODdjYWM2ODY4MDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.bLo_gxVxDsvosbDcwjXki9FDAONhbCXLbe_nwCpYZKM)
> 
> ### Screenshot from Websurfx with the `strong` and `span` tags included
> ![2024-01-01_08-03](https://private-user-images.githubusercontent.com/31008330/293525765-ae2bb279-804f-4085-9ada-4970c217d973.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQwODI0MTYsIm5iZiI6MTcwNDA4MjExNiwicGF0aCI6Ii8zMTAwODMzMC8yOTM1MjU3NjUtYWUyYmIyNzktODA0Zi00MDg1LTlhZGEtNDk3MGMyMTdkOTczLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAxVDA0MDgzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZhMWRlYjcyZDViNTgwMzc5NTRkNWRmY2Y5YzNkYjcyZjg4MjkzYzg1OWQwM2M3YWY5N2E1ZjM4ZmRkNzRjZjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jXp2U_MU4Bq0Eaod09LZEyzoW8ZaJpZW0cCmoNB1bAQ)

Ok after seeing the picture I agree with you now that those `spans` are kind of useless and needs to be removed but for the `strong` tags I would suggest keeping them because I think as mentioned earlier when they get rendered on the search page they will add an aesthetic style like showing the user what in the description of the site matched their query which can help provide a better user experience. :slightly_smiling_face: